File: 2021/09/18/q1/q1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public int countKDifference(int[] nums, int k) {
        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            for (int j = i+1; j < nums.length; j++) {
                if (Math.abs(nums[i]- nums[j]) == k) {
                    count++;
                }
            }
        }
        return count;
    }

    public int[] findOriginalArray(int[] changed) {
        if (changed.length % 2 != 0) {
            return new int[0];
        }
        Map<Integer, Integer> m = new HashMap<>();
        for (int i = 0; i < changed.length; i++) {
            m.computeIfAbsent(changed[i], k -> 1);
            m.computeIfPresent(changed[i], (k, v) -> v+1);
        }
        Arrays.sort(changed);
        List<Integer> ans = new ArrayList<>();
        for (int i = changed.length-1; i >= 0 ; i--) {
            if (changed[i]==0) {
                if (m.get(0) > 1) {
                    m.computeIfPresent(0, (k,v) -> v-2);
                    ans.add(0);
                }
                continue;
            }
            if (changed[i]%2 == 0) {
                int v = changed[i] /2 ;
                if (m.get(changed[i]) > 0 && m.get(v) > 0) {
                    m.computeIfPresent(changed[i], (k, val) -> val-1);
                    m.computeIfPresent(v, (k, val) -> val-1);
                    ans.add(v);
                }
            }
        }
        return ans.size() == changed.length/2 ? (int[]) ans.stream().mapToInt(i->i).toArray() : new int[0];
    }
}

File: 2021/09/18/q1/q1.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(findOriginalArray([]int{0, 0, 0, 0}))
}

func countKDifference(nums []int, k int) int {
	count := 0
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			if abs(nums[i]-nums[j]) == k {
				count++
			}
		}
	}
	return count
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func findOriginalArray(changed []int) []int {
	if len(changed)%2 != 0 {
		return []int{}
	}
	m := make(map[int]int)
	for _, v := range changed {
		m[v]++
	}
	sort.Ints(changed)
	ans := make([]int, 0)
	for i := len(changed) - 1; i >= 0; i-- {
		if changed[i] == 0 {
			if m[0] > 1 {
				m[0] -= 2
				ans = append(ans, 0)
			}
			continue
		}
		if changed[i]%2 == 0 {
			v := changed[i] / 2
			if m[v] > 0 && m[changed[i]] > 0 {
				m[v]--
				m[changed[i]]--
				ans = append(ans, v)
			}
		}
	}
	if len(ans) == len(changed)/2 {
		return ans
	}
	return []int{}
}
